daffodil-propgen/src/main/scala/UpdateEclipseClasspaths.scala

```diff
@@ -20,37 +20,33 @@ import java.io.PrintStream
 import scala.language.reflectiveCalls
 
 /**
- * Run this app if you change dependencies on libraries in the build.sbt file.
- * First do a 'sbt updateClassifiers compile'. Then run this.
- * <p>
- * This will edit all the eclipse-project/.classpath files to reference the new
- * versions, setting up the source and javadoc links as well for eclipse.
- *
- * Then you have to refresh all the daffodil eclipse projects in eclipse, and do a clean
- * and build.
+ * After 'sbt eclipse' creates the projects, you still have to update every .classpath file
+ * so that all but daffodil-macro-lib have a dependency on daffodil-macro-lib.
  */
 
-object UpdateDaffodilClasspaths extends UpdateEclipseClasspaths with App {
+object UpdateDaffodilClasspaths extends App {
```

Comment:
It would be nice if this file were turned into an SBT task instead of something that needs to be manually run. Something like "sbt updateEclipseClasspath", then instructions just become run ``sbt eclipse fixEclipseClasspath``.

It would also mean you wouldn't need to have DAFFODIL_HOME defined, since sbt knows where DAFFODIL_HOME is (i.e. baseDirectory.value). There's also an sbt variable for lib_managed (managedDirectory.value).  You mgith even be able to get rid of things like listFiles(), since I think you can get all submodules baseDirectory + "/.classpath". Not a big deal, but simplifies things a bit.

Reply:
Good idea. If I make this an sbt task it just lives in build.sbt or project/build.scala and I can make a task eclipseSetup which depends on eclipse and does the fixup. 

I will try this. I think DAFFODIL_HOME requirement goes away, which is nice. Much of the rest of the setup described on the wiki is still needed, but it does get simpler.

Reply:
I like the eclipseSetup + dependency idea. Note that this does mean that we need to add the sbt eclipse plugin to plugins.sbt. I don't think that's a big deal.

I'd also recommend putting the bulk of this code in project/SetupEclipse.scala or something and then reference it from build.sbt, similar to how we do Rat.scala and Dependencies.scala. Keeps things a bit more organized and separate.

Reply:
I have converted UpdateEclipseClasspaths to an sbt task.

I was unable to figure out how to invoke an sbt command like 'sbt eclipse' as a task. There are no taskKeys defined in that plugin (from examining its source), so it appears to have been created with command line usage in mind vs. automated maintenance of eclipse files when dependencies change over time. 

But this is still pretty easy. You need the sbtEclipse plugin (in your ~/.sbt/1.0/plugins.sbt)
You just invoke sbt eclipse updateEclipseClasspaths.

I still am unable to get past scala 2.12 here.  I get:
{code}
Error in Scala compiler: null
Project: "daffodil-cli" in scope: "main" not built due to errors in dependent project(s): daffodil-japi. Root error(s): Error in Scala compiler: null;SBT builder crashed while compiling. The error message is 'null'. Check Error Log for details.
SBT builder crashed while compiling. The error message is 'null'. Check Error Log for details.
{code}
In the error log I find a java.lang.IncompatibleClassChangeError.
The stacktrace is below. 
{code}
java.lang.IncompatibleClassChangeError
	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:488)
	at java.lang.StringBuilder.append(StringBuilder.java:166)
	at com.typesafe.genjavadoc.JavaSig.rec$1(JavaSig.scala:64)
	at com.typesafe.genjavadoc.JavaSig.fullNameInSig$1(JavaSig.scala:67)
	at com.typesafe.genjavadoc.JavaSig.toJava$1(JavaSig.scala:168)
	at com.typesafe.genjavadoc.JavaSig.js(JavaSig.scala:178)
	at com.typesafe.genjavadoc.JavaSig.js$(JavaSig.scala:9)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.js(Plugin.scala:93)
	at com.typesafe.genjavadoc.AST$MethodInfo$.apply(AST.scala:138)
	at com.typesafe.genjavadoc.BasicTransform.$anonfun$addMethod$1(BasicTransform.scala:151)
	at scala.Option.map(Option.scala:146)
	at com.typesafe.genjavadoc.BasicTransform.addMethod(BasicTransform.scala:151)
	at com.typesafe.genjavadoc.BasicTransform.newTransform(BasicTransform.scala:110)
	at com.typesafe.genjavadoc.BasicTransform.newTransform$(BasicTransform.scala:48)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.newTransform(Plugin.scala:93)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transform(Plugin.scala:103)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transform(Plugin.scala:93)
	at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2589)
	at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
	at scala.reflect.internal.Trees.itransform(Trees.scala:1415)
	at scala.reflect.internal.Trees.itransform$(Trees.scala:1347)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.superTransform(Plugin.scala:102)
	at com.typesafe.genjavadoc.BasicTransform.newTransform(BasicTransform.scala:115)
	at com.typesafe.genjavadoc.BasicTransform.newTransform$(BasicTransform.scala:48)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.newTransform(Plugin.scala:93)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transform(Plugin.scala:103)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transform(Plugin.scala:93)
	at scala.reflect.api.Trees$Transformer.transformTemplate(Trees.scala:2563)
	at scala.reflect.internal.Trees.$anonfun$itransform$4(Trees.scala:1419)
	at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
	at scala.reflect.internal.Trees.itransform(Trees.scala:1418)
	at scala.reflect.internal.Trees.itransform$(Trees.scala:1347)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.superTransform(Plugin.scala:102)
	at com.typesafe.genjavadoc.BasicTransform.$anonfun$newTransform$3(BasicTransform.scala:90)
	at com.typesafe.genjavadoc.BasicTransform.withClass(BasicTransform.scala:136)
	at com.typesafe.genjavadoc.BasicTransform.newTransform(BasicTransform.scala:90)
	at com.typesafe.genjavadoc.BasicTransform.newTransform$(BasicTransform.scala:48)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.newTransform(Plugin.scala:93)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transform(Plugin.scala:103)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transform(Plugin.scala:93)
	at scala.reflect.api.Trees$Transformer.$anonfun$transformStats$1(Trees.scala:2589)
	at scala.reflect.api.Trees$Transformer.transformStats(Trees.scala:2587)
	at scala.reflect.internal.Trees.$anonfun$itransform$7(Trees.scala:1437)
	at scala.reflect.api.Trees$Transformer.atOwner(Trees.scala:2600)
	at scala.reflect.internal.Trees.itransform(Trees.scala:1437)
	at scala.reflect.internal.Trees.itransform$(Trees.scala:1347)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.internal.SymbolTable.itransform(SymbolTable.scala:16)
	at scala.reflect.api.Trees$Transformer.transform(Trees.scala:2555)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.superTransform(Plugin.scala:102)
	at com.typesafe.genjavadoc.BasicTransform.newTransform(BasicTransform.scala:114)
	at com.typesafe.genjavadoc.BasicTransform.newTransform$(BasicTransform.scala:48)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.newTransform(Plugin.scala:93)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transform(Plugin.scala:103)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transform(Plugin.scala:93)
	at scala.tools.nsc.ast.Trees$Transformer.transformUnit(Trees.scala:140)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.superTransformUnit(Plugin.scala:101)
	at com.typesafe.genjavadoc.BasicTransform.newTransformUnit(BasicTransform.scala:19)
	at com.typesafe.genjavadoc.BasicTransform.newTransformUnit$(BasicTransform.scala:18)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.newTransformUnit(Plugin.scala:93)
	at com.typesafe.genjavadoc.GenJavadocPlugin$MyComponent$GenJavadocTransformer.transformUnit(Plugin.scala:104)
	at scala.tools.nsc.transform.Transform$Phase.apply(Transform.scala:30)
	at scala.tools.nsc.Global$GlobalPhase.$anonfun$applyPhase$1(Global.scala:426)
	at scala.tools.nsc.Global$GlobalPhase.applyPhase(Global.scala:419)
	at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1(Global.scala:390)
	at scala.tools.nsc.Global$GlobalPhase.$anonfun$run$1$adapted(Global.scala:390)
	at scala.collection.Iterator.foreach(Iterator.scala:929)
	at scala.collection.Iterator.foreach$(Iterator.scala:929)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1417)
	at scala.tools.nsc.Global$GlobalPhase.run(Global.scala:390)
	at scala.tools.nsc.Global$Run.compileUnitsInternal(Global.scala:1431)
	at scala.tools.nsc.Global$Run.compileUnits(Global.scala:1416)
	at scala.tools.nsc.Global$Run.compileSources(Global.scala:1412)
	at scala.tools.nsc.Global$Run.compile(Global.scala:1515)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:131)
	at xsbt.CachedCompiler0.run(CompilerInterface.scala:106)
	at xsbt.CompilerInterface.run(CompilerInterface.scala:32)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sbt.internal.inc.AnalyzingCompiler.call(AnalyzingCompiler.scala:237)
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:111)
	at sbt.internal.inc.AnalyzingCompiler.compile(AnalyzingCompiler.scala:90)
	at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compile$3(MixedAnalyzingCompiler.scala:81)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:12)
	at sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:132)
	at sbt.internal.inc.MixedAnalyzingCompiler.compileScala$1(MixedAnalyzingCompiler.scala:72)
	at sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:115)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:305)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:305)
	at sbt.internal.inc.Incremental$.doCompile(Incremental.scala:101)
	at sbt.internal.inc.Incremental$.$anonfun$compile$4(Incremental.scala:82)
	at sbt.internal.inc.IncrementalCommon.recompileClasses(IncrementalCommon.scala:118)
	at sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:64)
	at sbt.internal.inc.Incremental$.$anonfun$compile$3(Incremental.scala:84)
	at sbt.internal.inc.Incremental$.manageClassfiles(Incremental.scala:129)
	at sbt.internal.inc.Incremental$.compile(Incremental.scala:75)
	at sbt.internal.inc.IncrementalCompile$.apply(Compile.scala:70)
	at sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:309)
	at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:267)
	at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:158)
	at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:237)
	at sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:145)
	at org.scalaide.core.internal.builder.zinc.CachingCompiler.compile(CachingCompiler.scala:56)
	at org.scalaide.core.internal.builder.zinc.EclipseSbtBuildManager.aggressiveCompile(EclipseSbtBuildManager.scala:219)
	at org.scalaide.core.internal.builder.zinc.EclipseSbtBuildManager.runCompiler(EclipseSbtBuildManager.scala:152)
	at org.scalaide.core.internal.builder.zinc.EclipseSbtBuildManager.update(EclipseSbtBuildManager.scala:134)
	at org.scalaide.core.internal.builder.zinc.EclipseSbtBuildManager.build(EclipseSbtBuildManager.scala:85)
	at org.scalaide.core.internal.project.scopes.BuildScopeUnit.build(BuildScopeUnit.scala:72)
	at org.scalaide.core.internal.project.SbtScopesBuildManager.$anonfun$build$3(SbtScopesBuildManager.scala:53)
	at org.scalaide.core.internal.project.SbtScopesBuildManager.$anonfun$build$3$adapted(SbtScopesBuildManager.scala:50)
	at scala.collection.immutable.List.foreach(List.scala:389)
	at org.scalaide.core.internal.project.SbtScopesBuildManager.build(SbtScopesBuildManager.scala:50)
	at org.scalaide.core.internal.project.ScalaProject.build(ScalaProject.scala:596)
	at org.scalaide.core.internal.builder.ScalaBuilder.build(ScalaBuilder.scala:111)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:735)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:206)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:246)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:301)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:304)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:360)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:383)
	at org.eclipse.core.internal.resources.Workspace.buildInternal(Workspace.java:487)
	at org.eclipse.core.internal.resources.Workspace.build(Workspace.java:399)
	at org.eclipse.ui.actions.GlobalBuildAction$1.run(GlobalBuildAction.java:177)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:56)
{code}